presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java

```diff
@@ -0,0 +1,282 @@
+                if (value.equals(assigmentKeyExpression)) {
+                    value = assignmentValue;
+                }
+                if (valueList.equals(assigmentKeyExpression)) {
+                    valueList = assignmentValue;
+                }
+                inPredicate = new InPredicate(value, valueList);
+            }
+            return inPredicate;
+        }
+
+        private ProjectNode tryAppendProjection(ProjectNode node, Symbol symbol)
+        {
+            if (node.getOutputSymbols().contains(symbol)) {
+                return node;
+            }
+            else if (node.getSource().getOutputSymbols().contains(symbol)) {
+                ImmutableMap.Builder<Symbol, Expression> builder = ImmutableMap.builder();
+                builder.putAll(node.getAssignments());
+                builder.put(symbol, symbol.toQualifiedNameReference());
+                return new ProjectNode(node.getId(), node.getSource(), builder.build());
+            }
+            else {
+                return node;
+            }
+        }
+
+        @Override
+        public PlanNode visitApply(ApplyNode node, RewriteContext<InsertSemiJoinContext> context)
+        {
+            InPredicate inPredicate = context.get().peekInPredicate();
+            if (inPredicate.getValueList() instanceof QualifiedNameReference) {
+                Symbol value = asSymbol(inPredicate.getValue());
+                Symbol valueList = asSymbol(inPredicate.getValueList());
+                if (node.getInput().getOutputSymbols().contains(value) && node.getSubquery().getOutputSymbols().contains(valueList)) {
+                    Symbol semijoinResult = symbolAllocator.newSymbol("semijoin_result", BOOLEAN);
+                    context.get().setInPredicateSymbol(semijoinResult);
+                    return new SemiJoinNode(idAllocator.getNextId(),
+                            node.getInput(),
+                            node.getSubquery(),
+                            value,
+                            valueList,
+                            semijoinResult,
+                            Optional.empty(),
+                            Optional.empty()
+                    );
+                }
+                else {
+                    return new ApplyNode(node.getId(),
+                            context.rewrite(node.getInput(), context.get()),
+                            node.getSubquery(),
+                            node.getCorrelation());
+                }
+            }
+            else {
+                return super.visitApply(node, context);
+            }
+        }
+
+        private Symbol asSymbol(Expression expression)
+        {
+            checkState(expression instanceof QualifiedNameReference);
+            return Symbol.fromQualifiedName(((QualifiedNameReference) expression).getName());
+        }
+    }
+
+    private class InsertSemiJoinContext
+    {
+        private LinkedList<InPredicate> inPredicates = new LinkedList<>();
```

Comment:
Or just replace with two fields: `originalInPredicate`, and `intermediateInPredicate?` as we talked.